First example (#200)
ZIB-IOL/FrankWolfe.jl: docs/src/examples.md

# Examples

> In this example, we project a random point onto a ``\ell_1``-norm ball with the basic Frank-Wolfe algorithm using

Reviewer: ```suggestion
In this example, we project a random point onto a probability simplex with the Frank-Wolfe algorithm using
```

> either the specialized LMO defined in the package or a generic LP formulation using MathOptInterface.jl and

Reviewer: ```suggestion
either the specialized LMO defined in the package or a generic LP formulation using MathOptInterface.jl (MOI) and
```

> either the specialized LMO defined in the package or a generic LP formulation using MathOptInterface.jl (MOI) and
> GLPK as underlying LP solver.

Reviewer: ```suggestion
either the specialized LMO defined in the package or a generic LP formulation using `MathOptInterface.jl` (MOI) and
`GLPK` as underlying LP solver.
```

function grad!(storage, x)
    @. storage = 2 * (x - xp)
    return nothing
end

lmo_radius = 2.5
lmo = FrankWolfe.LpNormLMO{Float64,1}(lmo_radius)

x00 = FrankWolfe.compute_extreme_point(lmo, zeros(n))
gradient = collect(x00)

x_lmo, v, primal, dual_gap, trajectory_lmo = FrankWolfe.frank_wolfe(
    f,
    grad!,
    lmo,
    collect(copy(x00)),
    max_iteration=k,
    line_search=FrankWolfe.Shortstep(),
    L=2,
    print_iter=k / 10,
    emphasis=FrankWolfe.memory,
    verbose=true,
    trajectory=true,
);
```
Create a MathOptInterface Optimizer and build the same linear constraints:
```@example 1
o = GLPK.Optimizer()
x = MOI.add_variables(o, n)

for xi in x
    MOI.add_constraint(o, xi, MOI.GreaterThan(0.0))
end

MOI.add_constraint(
    o,
    MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(1.0, x), 0.0),
    MOI.EqualTo(lmo_radius),
)

lmo_moi = FrankWolfe.MathOptLMO(o)

x, v, primal, dual_gap, trajectory_moi = FrankWolfe.frank_wolfe(
    f,
    grad!,
    lmo_moi,
    collect(copy(x00)),
    max_iteration=k,
    line_search=FrankWolfe.Shortstep(),
    L=2,
    print_iter=k / 10,
    emphasis=FrankWolfe.memory,
    verbose=true,
    trajectory=true,
);
```

> Alternatively, we can use one of the modelling interfaces based on MOI to formulate the LP. The following example builds the same set of constraints using JuMP:

Reviewer: ```suggestion
Alternatively, we can use one of the modelling interfaces based on `MOI` to formulate the LP. The following example builds the same set of constraints using `JuMP`:
```

> Alternatively, we can use one of the modelling interfaces based on MOI to formulate the LP. The following example builds the same set of constraints using JuMP:

Reviewer: ```suggestion
Alternatively, we can use one of the modelling interfaces based on `MOI` to formulate the LP. The following example builds the same set of constraints using `JuMP`:
```

# Examples

> In this example, we project a random point onto a ``l_1``-norm ball with the basic Frank-Wolfe algorithm using

Reviewer: ```suggestion
In this example, we project a random point onto a ``\ell_1``-norm ball with the basic Frank-Wolfe algorithm using
```

either the specialized LMO defined in the package or a generic LP formulation using MathOptInterface.jl and
GLPK as underlying LP solver.

> It can be found as example 4.4 [here](https://arxiv.org/pdf/2104.06675.pdf).

Reviewer: ```suggestion
It can be found as Example 4.4 [in the paper](https://arxiv.org/abs/2104.06675).
```

    @. storage = 2 * (x - xp)
    return nothing
end

lmo_radius = 2.5
lmo = FrankWolfe.LpNormLMO{Float64,1}(lmo_radius)

x00 = FrankWolfe.compute_extreme_point(lmo, zeros(n))
gradient = collect(x00)

x_lmo, v, primal, dual_gap, trajectory_lmo = FrankWolfe.frank_wolfe(
    f,
    grad!,
    lmo,
    collect(copy(x00)),
    max_iteration=k,
    line_search=FrankWolfe.Shortstep(),
    L=2,
    print_iter=k / 10,
    emphasis=FrankWolfe.memory,
    verbose=true,
    trajectory=true,
);
```
Create a MathOptInterface Optimizer and build the same linear constraints:
```@example 1
o = GLPK.Optimizer()
x = MOI.add_variables(o, n)

# x_i ≥ 0
for xi in x
    MOI.add_constraint(o, xi, MOI.GreaterThan(0.0))
end
# ∑ x_i == 1
MOI.add_constraint(
    o,
    MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(1.0, x), 0.0),
    MOI.EqualTo(lmo_radius),
)

lmo_moi = FrankWolfe.MathOptLMO(o)

x, v, primal, dual_gap, trajectory_moi = FrankWolfe.frank_wolfe(
    f,
    grad!,
    lmo_moi,
    collect(copy(x00)),
    max_iteration=k,
    line_search=FrankWolfe.Shortstep(),
    L=2,
    print_iter=k / 10,
    emphasis=FrankWolfe.memory,
    verbose=true,
    trajectory=true,
);
```

> Formulate the LP using JuMP:

Reviewer: ```suggestion
Alternatively, we can use one of the modelling interfaces based on MOI to formulate the LP. The following example builds the same set of constraints using JuMP:
```

```@example 1
m = JuMP.Model(GLPK.Optimizer)
@variable(m, y[1:n] ≥ 0)
# ∑ x_i == 1
@constraint(m, sum(y) == lmo_radius)

lmo_jump = FrankWolfe.MathOptLMO(m.moi_backend)

x, v, primal, dual_gap, trajectory_jump = FrankWolfe.frank_wolfe(
    f,
    grad!,
    lmo_jump,
    collect(copy(x00)),
    max_iteration=k,
    line_search=FrankWolfe.Shortstep(),
    L=2,
    print_iter=k / 10,
    emphasis=FrankWolfe.memory,
    verbose=true,
    trajectory=true,
);

x_lmo, v, primal, dual_gap, trajectory_lmo_blas = FrankWolfe.frank_wolfe(
    f,
    grad!,
    lmo,
    x00,
    max_iteration=k,
    line_search=FrankWolfe.Shortstep(),
    L=2,
    print_iter=k / 10,
    emphasis=FrankWolfe.blas,
    verbose=true,
    trajectory=true,
);

x, v, primal, dual_gap, trajectory_jump_blas = FrankWolfe.frank_wolfe(
    f,
    grad!,
    lmo_jump,
    x00,
    max_iteration=k,
    line_search=FrankWolfe.Shortstep(),
    L=2,
    print_iter=k / 10,
    emphasis=FrankWolfe.blas,
    verbose=true,
    trajectory=true,
);

> # Defined the x-axis for the series, when plotting in terms of iterations.

Reviewer: remove the inbetween comments, here, they are not essential for the tutorial
